WordPress/openverse
# Query the filtered indexes

## Fixes



Fixes #1198 by @sarayourfriend 

## Description


Add a new setting `ENABLE_FILTERED_INDEX_QUERIES` and a new search parameter `unstable__include_sensitive_results`. Query the filtered index when the setting is enabled and the query parameter is false. Finally, modifies the abstract mature and deleted media classes to accommodate the filtered indexes. This was an unexpected need that was missed during implementation planning.


## Testing Instructions

Checkout this branch and test the default behaviour which does not query the filtered indexes. Run `just api/up` to get the API running.

The following requests verify one of the mock sensitive terms used for the local filtered index creation, "bird". Please reference `ingestion_server/static/mock_sensitive_terms.txt` for other terms to test with.

These requests should include results that reference birds:

- http://localhost:50280/v1/images/?q=perching
- http://localhost:50280/v1/images/?q=bird

Now test this PR with the new setting enabled. Add the new setting to your `api/.env` file:

```
$ echo "ENABLE_FILTERED_INDEX_QUERIES=True" >> api/.env
```

If you already have the API running, you'll need to restart it for the setting to take effect:

```
$ just down
$ just api/up
```

The following requests should now all differ. Requests without the `include_sensitive_results` params should not include any results matching "bird":

- http://localhost:50280/v1/images/?q=perching
- http://localhost:50280/v1/images/?q=perching&unstable__include_sensitive_results=True
- http://localhost:50280/v1/images/?q=bird
    - **These results should be entirely empty**
- http://localhost:50280/v1/images/?q=bird&unstable__include_sensitive_results=True

We must also test that the existing mature filter for documents with `mature=true` in Elasticsearch still works.

From your local Django admin (http://localhost:50280/admin, username and password both are deploy), create an image report with the reason set to mature. I used the identifier `6d6d8c7d-9f86-4bb4-b332-cb039f798fbf` because it is the first result in the `perching` query so is easy to verify. Set the decision to "mature_filtered" so that the result gets update in Elasticsearch. Now run the query again and confirm that the identifier you made the report for does not appear in the filtered results. Confirm that the result does appear when the `include_sensitive_results` parameter is true.

Now do the same with another result but deindex it this time instead of marking it as mature. Confirm that it does not appear in _any_ searches, regardless of whether sensitive results are included.

## Checklist


- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
